### DISCUSSION TITLE
Setting up 2FA for FalconPy?

### DISCUSSION THREAD

**Discussion by cyberbeach33** [CrowdStrike/falconpy#1084] [Q&A] [answered]
Hi, Is there a document explaining how to setup 2FA for FalconPy (using okta/duo/etc)? Also, I've used PSFalcon in the past and used Microsoft Powershell SecretVault to store creds, is there a similar way to do auth for FalconPy? Also, if there a way to push auth to a phone when a user wants to run falconpy scripts? Thanks!

**Comment by cyberbeach33** [comment]
Additional comment... does anyone have a key management system they recommend that is locally hosted (in a file/encrypted) and requires auth to get new Client Token for use? Again, the "SecretVault" from MSFT in powershell works pretty well, but would love to have a python solution not requiring PS.

**Comment by jshcodes** [comment, marked as answer]
Hi cyberbeach33 -

Thanks for the questions! 😄 

> [!NOTE]
> While FalconPy maintains credentials and authentication state within the [Authentication Object](https://falconpy.io/Usage/Extensibility.html#falconinterface), it does not maintain information about how these credentials are provided or additional authentication factors used. This is intentional, and allows for more developer flexibility regarding these types of requirements.

Let’s start with credential storage.

### Storing credentials
There are _many_ examples of storing credentials using Python online, with none of them necessarily being the “right” or “wrong” answer. Each solution has unique advantages and disadvantages. Some solutions may align to the needs of your environment, others… maybe not so much.

#### FalconPy repository examples
We have a few examples demonstrating potential solutions here within this repo:

- [Local encryption](https://github.com/CrowdStrike/falconpy/blob/main/samples/authentication/aes_authentication.py) - Store credentials locally encrypted within a file or the environment. There is a discussion post related to this solution [here](https://github.com/CrowdStrike/falconpy/discussions/792). This example demonstrates AES/CBC but can be modified as needed.
- [AWS Parameter Store](https://github.com/CrowdStrike/falconpy/blob/main/samples/authentication/aws_parameter_store.py) - Store credentials within AWS Parameter Store.
- [Azure Key Vault](https://github.com/CrowdStrike/falconpy/blob/main/samples/authentication/azure_key_vault.py) - Store credentials within Azure Key Vault.
- [GCP Secrets Manager](https://github.com/CrowdStrike/falconpy/blob/main/samples/authentication/gcp_secrets_manager.py) - Store credentials within GCP Secrets Manager.

#### A few other possibilities
Additional options that leverage different vendors:

- [CyberArk](https://docs.cyberark.com/PAS/13.0/en/Content/WebServices/Implementing%20Privileged%20Account%20Security%20Web%20Services%20.htm) - Found a possible example [here](https://github.com/infamousjoeg/cyberark-examples/blob/master/python_example.py).
- [HashiCorp Vault](https://www.hashicorp.com/products/vault) - Simple example [here](https://github.com/hashicorp/vault-examples/blob/main/examples/_quick-start/python/example.py).
* [Okta](https://github.com/okta/okta-sdk-python) may be more complex depending on what we are trying to do. [Here](https://developer.okta.com/blog/2022/07/27/python-okta) is a simple flask project example that also talks about authentication and may speak to pieces of this ask.


### Two-factor authentication
As mentioned, FalconPy doesn’t handle this part of your application, but there are several potential solutions that would work in conjunction with a project leveraging FalconPy. Some of the available options provide SMS / phone interaction as an available second factor, others you may need to do more work to get there. All of these vendors provide documented SDKs for interacting with their APIs. These solutions could also be potentially mixed with the examples mentioned above.

> [!IMPORTANT]
> Depending on implementation, it may be trivial for your application users to circumvent these solutions if they have access to the Python source.

- [Auth0](https://github.com/auth0/auth0-python) - A writeup for working with Auth0 and Python can be found [here](https://auth0.com/blog/use-the-auth0-python-sdk-for-querying-and-storing-users-data/).
- [Duo](https://github.com/duosecurity/duo_client_python) - A writeup for working with Duo and Python can be found [here](https://duo.com/docs/authapi-guide).
- [Google Authenticator](https://www.geeksforgeeks.org/two-factor-authentication-using-google-authenticator-in-python/)
- [Plivo](https://www.plivo.com/docs/sms/use-cases/2-factor-authentication/python/) - Writeup documentation can be found [here](https://www.plivo.com/docs/sdk/server/python-sdk).
- [Twilio](https://github.com/twilio/twilio-python) / [Authy](https://github.com/twilio/authy-python) - [Blog post](https://www.twilio.com/blog/push-two-factor-authentication-python-twilio-authy) describing this solution.

**Reply by cyberbeach33** [reply]
Excellent. Thanks for the help!
